Add email notification on project/release removal

MVrachev opened the pull request:
Until now, when there are multiple contributors on a single
the project, if one of them deletes a release or the whole
project the other contributors don't get any notification,
which is problematic.

Connected with issue https://github.com/pypa/warehouse/issues/5714

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

di commented:
Hi @MVrachev, is this still a WIP? I don't see any changes here that actually use these new templates.

MVrachev commented:
I consciously split the implementation of the issue https://github.com/pypa/warehouse/issues/5714 into multiple pull requests.

Here I wanted to discuss with you the text in the templates themselves because it's my first time using Jinjra2.

There will be another pull request which will use those templates.

di commented:
Thanks for the explanation. Let's keep everything in one pull request, the maintainers have limited bandwidth and it's easier for us to review a single feature-complete PR.

If you have any questions about the implementation, feel free to ask in #5714.

MVrachev commented:
@di I have updated my pull request with the code using it.
I tested it for both package removal and package release removal.

MVrachev commented:
I will fix the issues with the CI these days and you can have a look again.

MVrachev commented:
@di I updated the unit tests in the pr.

The tests I touched are not failing but some others do, but it seems they are failing in the master branch too.

I think the pr is ready for a review. It's complete.

ewdurbin commented on warehouse/email/__init__.py:
```diff
@@ -201,6 +201,48 @@ def send_added_as_collaborator_email(request, user, *, submitter, project_name,
     return {"project": project_name, "submitter": submitter.username, "role": role}
 
 
+@_email("removed-package")
```
We should probably maintain consistency by calling this `removed-project`.

ewdurbin commented on warehouse/email/__init__.py:
```diff
@@ -201,6 +201,48 @@ def send_added_as_collaborator_email(request, user, *, submitter, project_name,
 
 
+@_email("removed-package")
+def send_removed_package_email(
```
Consistency: `send_removed_project_email`

ewdurbin commented on warehouse/email/__init__.py:
```diff
@@ -201,6 +201,48 @@ def send_added_as_collaborator_email(request, user, *, submitter, project_name,
+
+
+@_email("removed-package-release")
+def send_removed_package_release_email(
```
I think it'd be OK to drop the "package" part here:

```python
@_email("removed-release")
def send_removed_release_email(
```

ewdurbin replied:
Could the signature for this be reduced to:

```
(request, user, *, release, submitter_name, submitter_role, recipient_role)
```

`project_name` can be retrieved from `release.project.name` and `release_date` can be retrieved from `self.release.created`

ewdurbin commented on warehouse/templates/email/removed-package/subject.txt:
```diff
@@ -0,0 +1,19 @@
+{% extends "email/_base/subject.txt" %}
+
+{% block content %}
+  {% trans project=project %}The package {{ project }} has been deleted.{% endtrans %}
```
Again, for consistency I think we should have this be "The project {{ project }} has been deleted."

ewdurbin commented on warehouse/templates/email/removed-package/body.txt:
```diff
@@ -0,0 +1,29 @@
+
+{% block reason %}
+  {% trans recipient_role_descr=recipient_role_descr %}
+    You are receiving this because you are {{ recipient_role_descr }} of this package.
```
package => project

MVrachev commented:
I addressed your feedback @ewdurbin. 

I was wondering what word should I use when there is a deletion of a particular release file in a project.
Should I call it `removed_package` because according to https://pypi.org/help/#packages:

`A "file", also known as a "package", on PyPI is something that you can download and install. Because of different hardware, operating systems, and file formats, a release may have several files (packages), like an archive containing source code or a binary wheel.
`

ewdurbin commented:
> I addressed your feedback @ewdurbin.
> 
> I was wondering what word should I use when there is a deletion of a particular release file in a project.
> Should I call it `removed_package` because according to https://pypi.org/help/#packages:
> 
> `A "file", also known as a "package", on PyPI is something that you can download and install. Because of different hardware, operating systems, and file formats, a release may have several files (packages), like an archive containing source code or a binary wheel.`

Thanks for those changes @MVrachev!

This question actually does stir up the fact that we're only notifying if a Project or Release as whole is removed.

If notification for single file removal is added (which is probably a good idea), I'd suggest that we use something along the lines of:

`send_removed_release_file_email` or `send_removed_file_email`. At the verbose extreme... `send_removed_project_release_file_email` works... as it follows the whole "chain" down (Projects hold Releases, Releases hold Files).

I'm mostly sticking to the names that we use within the codebase/schema; see [Project](https://github.com/pypa/warehouse/blob/c4cb5896558bd67591ebf8dc7141fe98a06a076f/warehouse/packaging/models.py#L102), [Release](https://github.com/pypa/warehouse/blob/c4cb5896558bd67591ebf8dc7141fe98a06a076f/warehouse/packaging/models.py#L304), and [File](https://github.com/pypa/warehouse/blob/c4cb5896558bd67591ebf8dc7141fe98a06a076f/warehouse/packaging/models.py#L458).

Overall, package is just kind of a messy term. If anything this nudges me to consider working to shore up that FAQ entry :)

MVrachev commented:
Are there other changes required you want me to do @ewdurbin ?
If not, can we merge this because I want to continue my work on issue #5714 and add email notifications when:
- Send a email notification to all collaborators when a file from a version is deleted.
- Send a email notification to all collaborators when new version of the package is uploaded.

ewdurbin commented:
It looks like there are some issues with test coverage and listing, once those are resolved this is good to go, thanks @MVrachev

MVrachev commented:
@ewdurbin you were right.
Sorry that I didn't looked at that in the first place before asking for a review.

I have added multiple new unit tests in order to increase the code coverage and during that process I found some small bugs and problems.

I think it requires a new look at the new changes. I made a new commit with them.

MVrachev commented:
I have changed the role of the submitter to "owner" for `send_removed_project_email` and `send_removed_project_release_email` and now Travis CI blocks because of the lack of translations.

What I am suppose to do?

MVrachev commented:
I fixed the translations issues.
I don't plan any new changes on that pr, but I think that maybe it's a good idea for a new review because of some recent changes.

ewdurbin commented:
Thank you @MVrachev!

monperrus commented:
Hi all,
I'd like to receive emails for each release, but it's unclear what else I should add in `pyproject.toml`

```
[project]
name = "crawler-user-agents"

authors = [
  { name="Martin Monperrus", email="martin.monperrus@gnieh.org" },
]
```

Any help would be appreciated, thanks!

di commented:
Hi, emails are sent to user accounts on PyPI, not addresses listed in project metadata. Hope that helps!

monperrus commented:
Thanks @di for the clarification.

Are emails sent for removals only? Or can they also be sent for normal releases? (NPM does this and it's super useful)

di commented:
See #13234.

monperrus commented:
thanks.
